cmd/server/main.go

```diff
@@ -7,186 +7,305 @@ import (
+	if limit < 0 || limit > int64(len(products)) {
+		w.WriteHeader(400)
+		return
+	}
+
+	rows, err := db.Query("SELECT id, name, description, price FROM products_table LIMIT ?", limit)
 	if err != nil {
 		log.Fatal(err)
 	}
 	defer rows.Close()
 
-	type prodDB struct {
-		id          int
-		name        string
-		description string
-		price       int64
-	}
-	var prods []prodDB
 	for rows.Next() {
-		var p prodDB
-		if err := rows.Scan(&p.id, &p.name, &p.description, &p.price); err != nil {
+		var id int
+		var name string
+		var description string
+		var price float64
+		if err := rows.Scan(&id, &name, &description, &price); err != nil {
 			log.Fatal(err)
 		}
-		prods = append(prods, p)
+		products = append(products, product{ID: id, Name: name, Description: description, Price: price})
 	}
 
 	err = rows.Err()
 	if err != nil {
 		log.Fatal(err)
 	}
-	status := "status: up"
-	if err := db.Ping(); err != nil {
-		status = "status: down"
-	}
-	log.Println(status)
-	log.Println(prods)
-	return
-
-}
-
-func getProducts(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json")
 	if err := json.NewEncoder(w).Encode(products); err != nil {
-		log.Println("error", err)
-		w.WriteHeader(http.StatusInternalServerError)
+		w.WriteHeader(500)
+		return
 	}
 }
 
 func getProduct(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	params := mux.Vars(r)
-	for _, item := range products {
-		if strconv.Itoa(item.ID) == params["id"] {
-			if err := json.NewEncoder(w).Encode(item); err != nil {
-				log.Println("error", err)
-				w.WriteHeader(http.StatusInternalServerError)
-				return
-			}
+	paramsConvert, err := strconv.Atoi(params["id"])
+	if paramsConvert < 1 {
+		w.WriteHeader(404)
+		return
+	}
+	if err != nil {
+		w.WriteHeader(400)
+		return
+	}
+
+	var products product
+
+	rows, err := db.Query("SELECT id, name, description, price FROM products_table WHERE id = ?", params["id"])
+	if err != nil {
+		w.WriteHeader(400)
+		return
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var id int
+		var name string
+		var description string
+		var price float64
+		if err := rows.Scan(&id, &name, &description, &price); err != nil {
+			log.Fatal(err)
+		}
+		products = product{
+			ID:          id,
+			Name:        name,
+			Description: description,
+			Price:       price,
 		}
 	}
+
+	if products.ID == 0 {
+		w.WriteHeader(400)
+		return
+	}
+	err = rows.Err()
+	if err != nil {
+		log.Fatal(err)
+	}
+	if err := json.NewEncoder(w).Encode(products); err != nil {
+		w.WriteHeader(500)
+		return
+	}
+	w.WriteHeader(200)
 }
 
 func createProduct(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	var prod product
 	if err := json.NewDecoder(r.Body).Decode(&prod); err != nil {
-		log.Println("error", err)
-		w.WriteHeader(http.StatusInternalServerError)
+		w.WriteHeader(400)
+		return
+	}
+	if prod.Name == "" {
```

Comment:
Surely a product's name is not the only thing that needs validation. Might be good to also put all validations behind a `validate` method on `Product`.